src/common/storage/backends/s3/Client.js

```diff
@@ -0,0 +1,199 @@
+/* globals define */
+define([
+    '../StorageClient'
+], function (
+    StorageClient,
+) {
+    const S3Storage = function (id, name, logger, config = {}) {
+        StorageClient.apply(this, arguments);
+        this.bucketName = config.bucketName || 'deepforge';
+        this.endpoint = config.endpoint || 'http://localhost:80';
+        this.config = this.createS3Config(config);
+        this.s3Client = null;
+        this.ready = this.initialize();
+    };
+
+    S3Storage.prototype = Object.create(StorageClient.prototype);
+
+
+    S3Storage.prototype.initialize = async function () {
+        if (require.isBrowser) {
+            return new Promise((resolve, reject) => {
+                try {
+                    require(['aws-sdk-min'], () => {
+                        /* eslint-disable no-undef*/
+                        this.initializeS3Client(AWS);
+                        /* eslint-disable no-undef*/
+                        resolve();
+                    });
+                } catch (err) {
+                    reject(err);
+                }
+            });
+        } else {
+            const AWS = require.nodeRequire('aws-sdk');
+            this.initializeS3Client(AWS);
+        }
+    };
+
+    S3Storage.prototype.initializeS3Client = function (AWS) {
+        this.s3Client = new AWS.S3(this.config);
+        if (this.s3Client) {
+            promisifyMethod(this.s3Client, 'createBucket');
+            promisifyMethod(this.s3Client, 'getObject');
+            promisifyMethod(this.s3Client, 'putObject');
+            promisifyMethod(this.s3Client, 'deleteObject');
+            promisifyMethod(this.s3Client, 'headObject');
+            promisifyMethod(this.s3Client, 'listObjectsV2');
+        }
+    };
+
+    S3Storage.prototype.getS3Client = async function () {
+        await this.ready;
+        return this.s3Client;
```

Comment:
See one of my later comments about updates here. Basically, it would be nice to remove `this.s3Client`